[release/3.1] Stop building on Ubuntu 18.04

dougbu opened the pull request:
- use 20.04 instead

nit: Remove Ubuntu version from one CI job display name

dougbu committed:
[release/3.1] Stop building on Ubuntu 18.04
- use 20.04 instead

nit: Remove Ubuntu version from one CI job display name

dougbu commented:
@halter73 @Tratcher I bet PR validation here is going to fail similarly to how [my manual #20220830.17 build failed](https://dev.azure.com/dnceng/internal/_build/results?buildId=1978053&view=ms.vss-test-web.build-test-results-tab&runId=50604072&resultId=100853&paneView=debug): A number of tests (`TlsHandshakeRejectsTlsLessThan12`, `DevCertWithInvalidPrivateKeyProducesCustomWarning`, `ClientAttemptingToUseUnsupportedProtocolIsLoggedAsDebug`, and `DoesNotSupportTls10`) failed in both attempts there.

For example,
``` text
Error message
Assert.Throws() Failure\nExpected: typeof(System.IO.IOException)
Actual:   typeof(System.Security.Authentication.AuthenticationException): Authentication failed, see inner exception.
---- System.Security.Authentication.AuthenticationException : Authentication failed, see inner exception.
-------- Interop+OpenSsl+SslException : SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL.
------------ Interop+Crypto+OpenSslCryptographicException : error:141E70BF:SSL routines:tls_construct_client_hello:no protocols available

Stack trace
at System.Net.Security.SslStream.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
at System.Net.Security.SslStream.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
at System.Net.Security.SslStream.ProcessAuthentication(LazyAsyncResult lazyResult, CancellationToken cancellationToken)
at System.Net.Security.SslStream.BeginAuthenticateAsClient(SslClientAuthenticationOptions sslClientAuthenticationOptions, CancellationToken cancellationToken, AsyncCallback asyncCallback, Object asyncState)
at System.Net.Security.SslStream.BeginAuthenticateAsClient(String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation, AsyncCallback asyncCallback, Object asyncState)
at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__64_1(String arg1, X509CertificateCollection arg2, SslProtocols arg3, AsyncCallback callback, Object state)
at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2,TArg3](Func`6 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions)
at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2,TArg3](Func`6 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions)
at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2,TArg3](Func`6 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state)
at System.Net.Security.SslStream.AuthenticateAsClientAsync(String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation)
[at Microsoft.AspNetCore.Server.Kestrel.InMemory.FunctionalTests.HttpsConnectionMiddlewareTests.<>c__DisplayClass15_0.<<DoesNotSupportTls10>b__3>d.MoveNext() in /_/src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs:line 388](https://dev.azure.com/dnceng/internal/_git/23330b81-8bf6-4092-af7d-db8dfba063b8?path=%2F_%2Fsrc%2FServers%2FKestrel%2Ftest%2FInMemory.FunctionalTests%2FHttpsConnectionMiddlewareTests.cs&version=GBdougbu%2Fubuntu.20.04%2F3.1&_a=contents&line=388&lineEnd=389&lineStartColumn=1&lineEndColumn=1&lineStyle=plain)
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
at System.Net.Security.SslStream.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
at System.Net.Security.SslStream.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
at System.Net.Security.SslStream.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
at System.Net.Security.SslStream.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
at System.Net.Security.SslStream.ProcessAuthentication(LazyAsyncResult lazyResult, CancellationToken cancellationToken)
at System.Net.Security.SslStream.BeginAuthenticateAsClient(SslClientAuthenticationOptions sslClientAuthenticationOptions, CancellationToken cancellationToken, AsyncCallback asyncCallback, Object asyncState)
at System.Net.Security.SslStream.BeginAuthenticateAsClient(String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation, AsyncCallback asyncCallback, Object asyncState)
at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__64_1(String arg1, X509CertificateCollection arg2, SslProtocols arg3, AsyncCallback callback, Object state)
at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2,TArg3](Func`6 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions)
at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2,TArg3](Func`6 beg
```

In at least one later branch, `ClientAttemptingToUseUnsupportedProtocolIsLoggedAsDebug()` looks for _any_ `Exception`. Should I make the same change here❔

dougbu commented:
/cc @BrennanConroy because the `IOException` ➡️ `Exception` change came in from #18584.

dougbu commented:
I'm thinking at least part of what's went on in that PR related to making the `throw`n `Exception`s consistent across platforms, at least where OpenSSL is used. It may be the case that the runtime team didn't backport their changes to 3.1.x. Is that right @wfurt and @karelz❔

dougbu commented:
More information also available in https://dev.azure.com/dnceng/public/_build/results?buildId=1979965&view=ms.vss-test-web.build-test-results-tab&runId=50627370&paneView=debug

halter73 commented:
@dougbu is the "[release/2.1]" in the title a typo? This is targeting `release/3.1`.

I don't the change in exception type matters here much even in servicing. It would be weird to have a real app that relies on AuthenticateAsClientAsync throwing an IOException instead of an AuthenticationException in these kinds of scenarios. The tests are just overly specific.

wfurt commented:
I don't really remember @dougbu. 2.1 is out of support and 3.1 support ends in December. Running on Ubuntu 20.04 seems fine. 22.04 is whole different beast as it comes with OpenSSL 3.

dougbu commented:
> @dougbu is the "[release/2.1]" in the title a typo? This is targeting release/3.1.

Yeah, my bad. Updating the title&hellip;

> I don't think changing the exception type here matters much even in servicing

I'll make the change in my next iteration

> I don't really remember @dougbu. 2.1 is out of support and 3.1 support ends in December.

Thanks for thinking back. (Note everything used in ASP.NET 2.1.x applications remains fully supported on .NET Framework:tm:)

> Running on Ubuntu 20.04 seems fine. 22.04 is whole different beast as it comes with OpenSSL 3.

In addition, 20.04 should remain in support until April 2030 😀

dougbu commented:
@mthalman @crummel @lbussell which `dotnet-buildtools/prereqs` image should I use to replace `mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-cross-arm64-alpine-85814b7-20200327151156`❔

I can't find anything w/ `alpine` in the name on <https://mcr.microsoft.com/v2/dotnet-buildtools/prereqs/tags/list> between `ubuntu-18.04-cross-arm64-alpine-85814b7-20200327151156` (how nice, we were at the latest 18.04-based image 😀) and `ubuntu-22.04-cross-arm-alpine-20220830132559-b016fdd`. We'd like an Ubuntu 20.04-based image for our `Linux Musl ARM64` build job but I'm not sure how to search for what we need.

dougbu commented:
If it matters, we use the `mcr.microsoft.com/dotnet/runtime-deps:2.1-alpine` for our `Linux Musl x64` build. Should we therefore consider mcr.microsoft.com/dotnet/runtime-deps:3.1.28-alpine3.15-arm64v8` or the `alpine3.16` image❔ (I'm having trouble parsing the old Docker files to find which Alpine version we were getting.)

dougbu committed:
Skip `TlsHandshakeRejectsTlsLessThan12` more and expand `Assert.Throws<>()`
- skip `TlsHandshakeRejectsTlsLessThan12` because TLS1.0 is not available on Ubuntu 20.04
- for the other 3 tests, `IOException` was too specific for the platform switch

dougbu commented:
/azp run

mthalman commented:
> which `dotnet-buildtools/prereqs` image should I use to replace `mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-cross-arm64-alpine-85814b7-20200327151156`❔
> 
> I can't find anything w/ `alpine` in the name on https://mcr.microsoft.com/v2/dotnet-buildtools/prereqs/tags/list between `ubuntu-18.04-cross-arm64-alpine-85814b7-20200327151156` (how nice, we were at the latest 18.04-based image 😀) and `ubuntu-22.04-cross-arm-alpine-20220830132559-b016fdd`. We'd like an Ubuntu 20.04-based image for our `Linux Musl ARM64` build job but I'm not sure how to search for what we need.

If there isn't an existing one, the expectation is that teams will submit PRs to https://github.com/dotnet/dotnet-buildtools-prereqs-docker to produce images that meet their needs. Since this is just a move from 18.04 to 20.04, this might be as simple as copy-and-paste and update the version number.

dougbu commented:
[Internal test build #20220912.21](https://dev.azure.com/dnceng/internal/_build/results?buildId=1991227&view=results) was successful, except for the `DevCertWithInvalidPrivateKeyProducesCustomWarning` failure addressed in the last commit here.

dougbu commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs:
```diff
@@ -402,15 +402,15 @@ public async Task DevCertWithInvalidPrivateKeyProducesCustomWarning()
 
             await loggerProvider.FilterLogger.LogTcs.Task.DefaultTimeout();
-            Assert.Equal(3, loggerProvider.FilterLogger.LastEventId);
+            Assert.Equal(1, loggerProvider.FilterLogger.LastEventId);
```
Not exactly sure why this needed to change. But it lines up w/ tests in later branches

wtgodbe reviewed (approved):
Infra changes seem fine

dougbu commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs:
```diff
@@ -402,15 +402,15 @@ public async Task DevCertWithInvalidPrivateKeyProducesCustomWarning()
 
             await loggerProvider.FilterLogger.LogTcs.Task.DefaultTimeout();
-            Assert.Equal(3, loggerProvider.FilterLogger.LastEventId);
+            Assert.Equal(1, loggerProvider.FilterLogger.LastEventId);
             Assert.Equal(LogLevel.Error, loggerProvider.FilterLogger.LastLogLevel);
```
@Tratcher @javiercn @halter73 I think all of you have played in this area and I could use some help. This test (`DevCertWithInvalidPrivateKeyProducesCustomWarning`) no longer exists in later branches; it was removed as part of 8e1e81ae78e876eecad1b816093f0519682e205d but that relates to product changes that were never backported to 3.1.x. Would like to get it working here or throw the test away.

Right now, the test fails on macOS on the first line (413) because the `LastEventId` actually is `3`. On Linux (which was the platform that caused me to change line 413), the test fails on the second line because the `LastLogLevel` is actually `Debug`. The change here discrepancy to the upgrade to Ubuntu 20.04; the test was working consistently between Ubuntu 18.04 and macOS before the upgrade.

I think my choices are to delete the test,
```suggestion
```
,
```suggestion
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                Assert.Equal(1, loggerProvider.FilterLogger.LastEventId);
                Assert.Equal(LogLevel.Debug, loggerProvider.FilterLogger.LastLogLevel);
            } else {
                Assert.Equal(3, loggerProvider.FilterLogger.LastEventId);
                Assert.Equal(LogLevel.Error, loggerProvider.FilterLogger.LastLogLevel);
            }
```
, or perhaps find a way to make `HttpsConnectionFilterLogger` ignore `Debug` events after it's seen an `Error`.

Thoughts❔

halter73 replied:
> , or perhaps find a way to make `HttpsConnectionFilterLogger` ignore `Debug` events after it's seen an `Error`.

This could make sense if the issue is just that a debug log is no occurring after the error. What are the full logs for the failing and passing versions of this tests? This is using `LoggedTest`, so we should have these, right?

wtgodbe replied:
I'm not seeing anything other than the stack tract that just points to the Assert.Equal failure, unfortunately. There are test .xml/.html files uploaded, but they don't contain any extra info

dougbu replied:
Not sure where the macOS test logs got lost but https://dev.azure.com/dnceng-public/_apis/resources/Containers/28570718/Linux_Test_Logs?itemPath=Linux_Test_Logs%2FInMemory.FunctionalTests%2Fnetcoreapp3.1%2FHttpsTests%2FDevCertWithInvalidPrivateKeyProducesCustomWarning.log at least shows what happened on Ubuntu 20.04:

``` text
[0.233s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "0HMKM2A62STN9" accepted.
[0.823s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "0HMKM2A62STN9" started.
[0.825s] [Microsoft.AspNetCore.Server.Kestrel.Https.Internal.HttpsConnectionMiddleware] [Debug] Failed to authenticate HTTPS connection.
System.IO.IOException: Authentication failed because the remote party has closed the transport stream.
   at System.Net.Security.SslStream.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStream.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStream.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslStream.ProcessAuthentication(LazyAsyncResult lazyResult, CancellationToken cancellationToken)
   at System.Net.Security.SslStream.BeginAuthenticateAsServer(SslServerAuthenticationOptions sslServerAuthenticationOptions, CancellationToken cancellationToken, AsyncCallback asyncCallback, Object asyncState)
   at System.Net.Security.SslStream.<>c.<AuthenticateAsServerAsync>b__69_0(SslServerAuthenticationOptions arg1, CancellationToken arg2, AsyncCallback callback, Object state)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2](Func`5 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, Object state, TaskCreationOptions creationOptions)
   at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2](Func`5 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, Object state, TaskCreationOptions creationOptions)
   at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2](Func`5 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, Object state)
   at System.Net.Security.SslStream.AuthenticateAsServerAsync(SslServerAuthenticationOptions sslServerAuthenticationOptions, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Https.Internal.HttpsConnectionMiddleware.InnerOnConnectionAsync(ConnectionContext context) in /_/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:line 204
[0.838s] [Microsoft.AspNetCore.Server.Kestrel] [Debug] Connection id "0HMKM2A62STN9" stopped.
[1.334s] [TestLifetime] [Information] Finished test DevCertWithInvalidPrivateKeyProducesCustomWarning in 1.3332518s
```

javiercn replied:
If we threw the test away in later versions, I am ok with removing it here to unblock things. We switched most of our coverage to CTI.

dougbu replied:
done as you suggest @javiercn

dougbu committed:
Remove `DevCertWithInvalidPrivateKeyProducesCustomWarning` test
- test does not behave consistently across platforms
- test already removed in later branches
  - testing in this area has moved to CTI (manual testing)
